src/arrays.jl

```diff
@@ -0,0 +1,388 @@
+using SymbolicUtils
+using StaticArrays
+import Base: eltype, length, ndims, size, axes, eachindex
+
+### Store Shape as a metadata in Term{<:AbstractArray} objects
+struct ArrayShapeCtx end
+
+
+#=
+ There are 2 types of array terms:
+   `ArrayOp{T<:AbstractArray}` and `Term{<:AbstractArray}`
+
+ - ArrayOp represents a Einstein-notation-inspired array operation.
+   it contains a field `term` which is a `Term` that represents the
+   operation that resulted in the `ArrayOp`.
+   I.e. will be `A*b` for the operation `(i,) => A[i,j] * b[j]` for example.
+   It can be `nothing` if not known.
+   - calling `shape` on an `ArrayOp` will return the shape of the array or `Unknown()`
+   - do not rely on the `symtype` or `shape` information of the `.term` when looking at an `ArrayOp`.
+     call `shape`, `symtype` and `ndims` directly on the `ArrayOp`.
+ - `Term{<:AbstractArray}`
+   - calling `shape` on it will return the shape of the array or `Unknown()`, and uses
+     the `ArrayShapeCtx` metadata context to store this.
+
+ The Array type parameter must contain the dimension.
+=#
+
+#### ArrayOp ####
+
+"""
+    ArrayOp(output_idx, expr, reduce)
+
+A tensor expression where `output_idx` are the output indices
+`expr`, is the tensor expression and `reduce` is the function
+used to reduce over contracted dimensions.
+"""
+struct ArrayOp{T<:AbstractArray} <: Symbolic{T}
+    output_idx # output indices
+    expr       # Used in pattern matching
+               # Useful to infer eltype
+    reduce
+    term
+    shape
+end
+
+function ArrayOp(T::Type, output_idx, expr, reduce, term)
+    sh = make_shape(output_idx, expr)
+    ArrayOp{T}(output_idx, expr, reduce, term, sh)
+end
+
+shape(aop::ArrayOp) = aop.shape
```

Comment:
This conflicts with `Distributions.shape`

Reply:
Fixed